test/integration/master/kms_transformation_test.go

```diff
@@ -140,22 +123,80 @@ func TestKMSProvider(t *testing.T) {
 	}
 
 	// Secrets should be un-enveloped on direct reads from Kube API Server.
-	s, err := test.restClient.CoreV1().Secrets(testNamespace).Get(testSecret, metav1.GetOptions{})
+	mustReadSecret(t, test, true)
+	test.printMetrics()
+}
+
+// TestKMSFailureImpactOnReads validates that KMS Plugin will not be called for secrets that were cached by api-server\pkg\storage\cacher.go
+// Cacher.go (assuming sufficient memory) should intercept requests for secrets that were recently created or recently retrieved - thus
+// alleviating the need for caching within envelop.go.
+func TestKMSFailureImpactOnReads(t *testing.T) {
+	pluginMock, test := mustSetupTest(t)
+	defer pluginMock.cleanUp()
+	defer test.cleanUp()
+
+	pluginMock.grpcServer.GracefulStop()
+	// Giving time for mock plugin to shutdown
+	time.Sleep(2 * time.Second)
+
+	// Expect hit - cache was populated when the test secret was created (i.e. cacher.go has a cached version of the secret in plain text).
+	mustReadSecret(t, test, false)
+}
+
+func mustSetupTest(t *testing.T) (*base64Plugin, *transformTest) {
+	pluginMock, err := NewBase64Plugin()
```

Comment:
nit: for future reference, this should be called a fake, not a mock.

Reply:
I called it mock because it provide sensing capabilities; it exposes requests submitted by kube-apiserver to tests. 
However, I do see your point - it is not a true mock since it provides implementation of the functionality under test. 
Will rename on a separate PR.